syncer/syncer.go

```diff
@@ -163,31 +177,61 @@ func (s *syncer) BulkSync(ctx context.Context, newBlockCallback func(*types.Bloc
 }
 
 func (s *syncer) WatchSync(ctx context.Context, callback func(*types.Block) bool) error {
+	localLatest := s.blockchain.Header().Number
+	skipList := make(map[peer.ID]bool)
+
 	// Loop until context is canceled
 	for {
+		select {
+		case <-s.newStatus:
+		case <-time.After(s.blockTimeout):
+			return errTimeout
+		}
+
+		// fetch local latest block
+		if header := s.blockchain.Header(); header != nil {
+			localLatest = header.Number
+		}
+
 		// pick one best peer
-		// peer := s.peerHeap.BestPeer()
+		bestPeer := s.peerMap.BestPeer(skipList)
+		if bestPeer == nil {
+			break
+		}
+
+		// if the bestPeer does not have a new block continue
+		if bestPeer.Number <= localLatest {
+			continue
+		}
 
 		// fetch block from the peer
+		lastNumber, isValidator, err := s.bulkSyncWithPeer(bestPeer.ID, callback)
+		if err != nil {
+			s.logger.Warn("failed to complete bulk sync with peer, try to next one", "peer ID", bestPeer.ID)
+		}
 
-		// verify the block
+		if err != nil || lastNumber < bestPeer.Number {
+			skipList[bestPeer.ID] = true
 
-		// write the block
+			// continue to next peer
+			continue
+		}
 
-		select {
-		case <-ctx.Done():
-			// context is canceled
+		if isValidator {
 			break
 		}
 	}
+
+	return nil
 }
 
-func (s *syncer) bulkSyncWithPeer(peerID peer.ID, newBlockCallback func(*types.Block)) (uint64, error) {
+func (s *syncer) bulkSyncWithPeer(peerID peer.ID, newBlockCallback func(*types.Block) bool) (uint64, bool, error) {
 	localLatest := s.blockchain.Header().Number
+	isValidator := false
```

Comment:
`isValidator` is not right name in syncer package

Reply:
Fixed, what do you think about this naming 3d198079c12bfe45e2b7f54e4aae7b9fddbea315